pipeline_dsl/resources/semver.py

```diff
@@ -1,47 +1,76 @@
-import os
+from pipeline_dsl.resources.resource import AbstractResource, ConcourseResource
 from pipeline_dsl.concourse import concourse_context
+from typing import Optional, Dict, Any
+from abc import ABC, abstractmethod
+import os
 
 
 class SemVerResource:
-    def __init__(self, name):
+    def __init__(self, name: str) -> None:
         self.name = name
         self.path = os.path.abspath(self.name)
 
-    def __str__(self):
+    def __str__(self) -> str:
         return self.name
 
-    def version(self):
+    def version(self) -> str:
         if concourse_context():
             with open(os.path.join(self.path, "version")) as f:
                 return f.read().strip()
         return "0.0.1"
 
 
-class SemVer:
-    def __init__(self, source, initial_version=None):
-        self.source = source
+class SemVer(AbstractResource[SemVerResource]):
+    def __init__(self, driver: "SemVerDriver", initial_version: str = None):
+        self.driver = driver
         self.initial_version = initial_version
 
-    def resource_type(self):
+    def resource_type(self) -> Optional[dict]:
         return None
 
-    def concourse(self, name):
-        result = {
-            "name": name,
-            "type": "semver",
-            "icon": "creation",
-            "source": self.source.concourse(),
-        }
+    def concourse(self, name: str) -> ConcourseResource:
+        result = ConcourseResource(
+            name=name,
+            type="semver",
+            icon="creation",
+            source=self.driver.concourse(),
+        )
         if self.initial_version is not None:
-            result["source"]["initial_version"] = self.initial_version
+            result.source["initial_version"] = self.initial_version
         return result
 
-    def get(self, name):
+    def get(self, name: str) -> SemVerResource:
         return SemVerResource(name)
 
 
-class SemVerGitDriver:
-    def __init__(self, uri, branch, file, private_key=None, username=None, password=None, git_user=None, depth=None, skip_ssl_verification=None, commit_message=None):
+class SemVerDriver(ABC):
+    @abstractmethod
+    def resource_type(self) -> Optional[Dict]:
+        pass
+
+    @abstractmethod
+    def concourse(self) -> Dict[str, Any]:
+        pass
+
+    @abstractmethod
+    def get(self) -> None:
+        pass
```

Comment:
I don't think we need the other methods

```suggestion
class SemVerDriver(ABC):
    @abstractmethod
    def concourse(self) -> Dict[str, Any]:
        pass
```

Reply:
You are right. I removed them in https://github.com/SAP/pipeline-dsl/pull/16/commits/f57e178bab3eee96fab194f35bbe1bc7788bac85.